NXDOC-2086: initial draft for tutorial on nuxeo actions menu element (#1424)
nuxeo/doc.nuxeo.com-content: src/nxdoc/web-ui/web-ui-tutorials/how-to-create-and-customize-nuxeo-actions-menu.md

---
title: 'HOWTO: Create and Customize a Nuxeo actions menu'
review:
    comment: ''
    date: '2020-03-25'
    status: ok
toc: true
details:
    howto:
        excerpt: Learn how to create and customize a Nuxeo actions menu.
        level: Advanced
        tool: code
        topics: Web UI
labels:
    - lts2017-ok
    - lts2019-ok
    - tutorial
    - nuxeo-elements
    - nuxeo-ui-elements
    - polymer
tree_item_index: 500
---

{{! excerpt}}

> In this tutorial you will learn how to create and customize a Nuxeo actions menu on your code.

Reviewer: I don't know if people know what a Nuxeo actions menu is. We should explain that it is a responsive menu that can hold actions (e.g. buttons), which will hide some of them behind a dropdown according to the available (or predefined space).

{{! /excerpt}}

## Create an actions menu:
We add a nuxeo actions menu element.

```
<nuxeo-actions-menu>
</nuxeo-actions-menu>
```

> We add the menu items, which should consist of actions.

Reviewer: Starting with "We" again makes it repetitive. "And then, we add..", for example, would be an option.

> You can add an existing action, for example:

Reviewer: No need to separate this name from the one above. We could try to make it a succinct paragraph.

```
<nuxeo-add-to-collection-button document="[[document]]">
</nuxeo-add-to-collection-button>
```

> Or you can customize your own.

Reviewer: I guess the idea here is not so much customizing, because you either have to create a new one or update the code to make it work with the actions meu.

## Customize your own action:

> Include the nuxeo-action-button-styles module in the element style:

Reviewer: This requires `import '@nuxeo/nuxeo-ui-elements/actions/nuxeo-action-button-styles.js';` as well.

> Include the nuxeo-action-button-styles module in the element style:

Reviewer: We could explain there that `nuxeo-action-button-styles` is a styles module that bring the styles require for the action to be properly represented.

> Include the nuxeo-action-button-styles module in the element style:

Reviewer: These instructions in general feel a little bit prescriptive "Include this", "add that". We should try to add a little bit more flare to the text and write in a way that feels a little bit less like imposing. For example, here "We can start by adding the ..."

---
title: 'HOWTO: Create a Nuxeo actions menu'
review:
    comment: ''
    date: '2020-03-25'
    status: ok
toc: true
details:
    howto:
        excerpt: Learn how to create a Nuxeo actions menu.
        level: Advanced
        tool: code
        topics: Web UI
labels:
    - lts2017-ok
    - lts2019-ok
    - tutorial
    - nuxeo-elements
    - nuxeo-ui-elements
    - polymer
tree_item_index: 500
---

{{! excerpt}}
In this tutorial you will learn how to create and customize a Nuxeo actions menu on your code. The actions menu is a responsive menu that can hold actions (e.g. buttons), which hides behind a dropdown according to the available (or predefined) space.
{{! /excerpt}}

{{!--     ### nx_asset ###
    path: /default-domain/workspaces/Product Management/Documentation/Documentation Screenshots/NXDOC/Master/HOWTO: Create a Nuxeo actions menu/nuxeo actions menu screenshot
    name: nuxeo-action-menu.png
    1.1.3#screenshot#up_to_date
--}}
![nuxeo actions menu screenshot](nx_asset://9c147cdc-311a-4190-b7d1-7d2a255216f2 ?w=300,border=true)

## Create an actions menu:

> First, we start by adding a nuxeo actions menu element.

Author: thanks! suggestion added

```
<nuxeo-actions-menu>
</nuxeo-actions-menu>
```

Then we add the menu items, which should consist of actions - such as a nuxeo defined action, for example:

```

> <nuxeo-add-to-collection-button document="[[document]]">

Reviewer: It would help to also see the nuxeo-actions-menu tags, so we have an idea of how the whole thing would look like.

Author: thanks! suggestion added

</nuxeo-add-to-collection-button>
```

> Or you can create your own.

Reviewer: We need a little bit more context here, to why one would have to create (or adapt) an action.
```suggestion
Or you can also create new or adapt your own actions, to make them work with `nuxeo-actions-menu`.
```

Author: thanks! suggestion added

> ## Create your own action:

Reviewer: ```suggestion
## Create (or adapt) your own actions:
```

Author: thanks! suggestion added

> First, we import the ***nuxeo-action-button-styles** styles module to make sure the action is styled properly.

Reviewer: ```suggestion
First, we need to import the ***nuxeo-action-button-styles** styles module, to make sure the action is styled properly.
```

Author: thanks! suggestion added

```
import '@nuxeo/nuxeo-ui-elements/actions/nuxeo-action-button-styles.js';
```

And include the nuxeo-action-button-styles module in the element style.

```
<style include="nuxeo-action-button-styles">
  ...
</style>
```

Then, we add a showLabel boolean property (with false as the initial value).

```
showLabel: {
    type: Boolean,
    value: false,
},
```

> On the html, we add a span element with class label, with your label bound to it, right next to your button, and bind it’s hidden attribute to !showLabel.

Reviewer: "On the html" is rather vague. We need to be more concrete here.

Also, saying thinks like "bind it’s hidden attribute to !showLabel" isn't more helpful than just looking at the code. We should rather explain why we are doing this.

Suggestion:
```suggestion
On the action's DOM, we need to add a span element with class `label`, with your label bound to it, right next to your button. The span should be hidden when `showLabel` is `false`, so that it is only displayed when the action is in the dropdown.
```

Author: thanks! suggestion added

```
<paper-icon-button id="myAction" icon="nuxeo:someicon" on-tap=”_doSomething”>
</paper-icon-button>
<span class="label" hidden$="[[!showLabel]]">
  [[_label]]
</span>
```

> Finally, we wrap the two elements above (the button and the label) in a new div with class action, and move your tap or click event listener to the div:

Reviewer 2: `Finally, we wrap the two elements above (the button and the label) in a new div with class action, and move your tap or click event listener to it`:

Author: thanks! suggestion added

```
<div class="action" on-tap="_doSomething">
    <paper-icon-button noink id="myAction" icon="nuxeo:someicon">
    </paper-icon-button>
    <span class="label" hidden$="[[!showLabel]]">
      [[_label]]
    </span>
</div>
```

For more examples, you can check the [Web UI’s migration commit](https://github.com/nuxeo/nuxeo-web-ui/commit/60d5b637ef093d8279447d550391e686771ba115).

## Styling

Nuxeo-actions-menu dropdown can be styled by targeting the classes **--nuxeo-action-button-label** and **--nuxeo-actions-menu-dropdown**.

For example, to add some padding to each item in the dropdown, we can do:

```
--nuxeo-action-button-label: {
  padding-right: 8px;
}
--nuxeo-actions-menu-dropdown: {
  padding: 0 8px;
}
```

## Customizing the number of actions in action menus.

As of [NXP-25146](https://jira.nuxeo.com/browse/NXP-25146), the max-with of these three menus can be customized in your themes with three css variables:

- --nuxeo-browser-actions-menu-max-width (for the document actions menu)
- --nuxeo-results-selection-actions-menu-max-width (for the result selection actions menu)
- --nuxeo-document-blob-actions-menu-max-width (for the blob actions menu)

> These variables supports all the values valid for the max-width property, so you’ll have to set them to fit the desired number of actions.

Reviewer 2: ```suggestion
These variables support all the values valid for the max-width property, so you’ll have to set them to fit the desired number of actions.
```

Author: thanks! suggestion added

    name: nuxeo-action-menu.png
    1.1.3#screenshot#up_to_date
--}}
![nuxeo actions menu screenshot](nx_asset://9c147cdc-311a-4190-b7d1-7d2a255216f2 ?w=300,border=true)

## Create an actions menu:
First, we start by adding a nuxeo actions menu element to our custom element/page.

```
<nuxeo-actions-menu>
</nuxeo-actions-menu>
```

Then we add the menu items, which should consist of actions - such as a nuxeo defined action, for example:

```
<nuxeo-actions-menu>
  <nuxeo-add-to-collection-button document="[[document]]">
  </nuxeo-add-to-collection-button>
</nuxeo-actions-menu>
```

Or you can also create new or adapt your own actions, to make them work with ***nuxeo-actions-menu***.

## Create (or adapt) your own actions:

First, we need import the ***nuxeo-action-button-styles** styles module, to make sure the action is styled properly.

```
import '@nuxeo/nuxeo-ui-elements/actions/nuxeo-action-button-styles.js';
```

And include the nuxeo-action-button-styles module in the element style.

```
<style include="nuxeo-action-button-styles">
  ...
</style>
```

> Then, we add a showLabel boolean property (with false as the initial value).

Reviewer: ```suggestion
Then, we must add a `showLabel` boolean property (with false as the initial value).
```

Author: applied accepted suggestion

```
showLabel: {
    type: Boolean,
    value: false,
},
```

On the action's DOM, we need to add a span element with class `label`, with your label bound to it, right next to your button. The span should be hidden when `showLabel` is `false`, so that it is only displayed when the action is in the dropdown.

```
<paper-icon-button id="myAction" icon="nuxeo:someicon" on-tap=”_doSomething”>
</paper-icon-button>
<span class="label" hidden$="[[!showLabel]]">
  [[_label]]
</span>
```

> Finally, we wrap the two elements above (the button and the label) in a new div with class action, and move your tap or click event listener to it:

Reviewer: ```suggestion
Finally, we wrap the two elements above (the button and the label) in a new div with class `action`, and move your tap or click event listener to it:
```

Author: applied accepted suggestion
